src/App.tsx

```diff
@@ -136,26 +189,50 @@ function App() {
     });
   }
 
+  const setText = useCallback((newText: string) => {
+    const model = editor?.getModel();
+    if (!model || !editor) return;
+    model.pushEditOperations(
+      editor.getSelections(),
+      [
+        {
+          range: model.getFullModelRange(),
+          text: newText,
+        }
+      ],
+      () => null
+    );
+    editor.setPosition({ column: 0, lineNumber: 0});
+  }, [editor]);
+
   function handleLoadSample() {
-    if (editor?.getModel()) {
-      const model = editor.getModel()!;
-      model.pushEditOperations(
-        editor.getSelections(),
-        [
-          {
-            range: model.getFullModelRange(),
-            text: rustpadRaw,
-          },
-        ],
-        () => null
-      );
-      editor.setPosition({ column: 0, lineNumber: 0 });
-      if (language !== "rust") {
-        handleChangeLanguage("rust");
-      }
+    setText(rustpadRaw);
+    if (language !== "rust") {
+      handleChangeLanguage("rust");
     }
   }
 
+  const handleUploadFile = useCallback(async (file: File) => {
+    const text = await file.text();
+    setText(text);
+    const newLanguage = getLanguage(file.name);
+    if (newLanguage !== language) handleChangeLanguage(newLanguage);
+  }, [language]);
```

Comment:
I think you're missing `setText` in the dependency list for `handleUploadFile()`.

In general, I would prefer if we avoided `useCallback` unless it was absolutely necessary for efficiency reasons (such as if if you're passing a callback to a subcomponent wrapped in `memo`). Right now this isn't actually improving the efficiency of the page, since the function is getting created every time anyway - it's perhaps even making the code slightly (but unmeasurably) slower.

More importantly though, `useCallback` creates hard-to-find bugs like this one when you're missing dependencies. Could we perhaps just remove this `useCallback` invocation and not worry about this?